cmd/kubeadm/app/cmd/join.go

```diff
@@ -289,7 +289,7 @@ func (j *Join) Run(out io.Writer) error {
 	}
 
 	// Write env file with flags for the kubelet to use. Also register taints
-	if err := kubeletphase.WriteKubeletDynamicEnvFile(&j.cfg.NodeRegistration, j.cfg.FeatureGates, true, kubeadmconstants.KubeletRunDirectory); err != nil {
+	if err := kubeletphase.WriteKubeletDynamicEnvFile(&j.cfg.NodeRegistration, j.cfg.FeatureGates, "", true, kubeadmconstants.KubeletRunDirectory); err != nil {
```

Comment:
`""` in case of `JoinConfiguration`?

Reply:
This should take `ClusterConfiguration` and then invoke `GetPauseImage` in `WriteKubeletDynamicEnvFile`. But for that to be done, first we need to merge #66219 .
If setup like this, `kubelet join` can try to use the wrong pause image, thus causing this PR to only partially fix the problem (for masters only).

Reply:
makes sense, waiting on #66219.
if `ClusterConfiguration` contains `ImageRepository` - great.

but the v1beta1 KEP says:
> `v1beta1` API is going to simplify how a Kubernetes administrator can specify custom images by removing `ImageRepository` and `UnifiedControlPlaneInmage` attributes with the the `image` attribute (simpler, more flexible, and consistent with the `image` attribute already used into the `etcd` object)

so if `ImageRepository` is gone from `v1beta1` things like `pause` won't have a custom image repository.

Reply:
I do think, that `ImageRepository` is going to be kept here. Even if it's removed, it's going to be in favour of having separate image fields in the config for all K8s components. This includes the `pause` image. Given the fact, that the pause image is going to be needed on worked nodes too, it's setting is going to be sitting in the `ClusterConfiguration`. Hence, no matter if we have `ImageRepository` or `PauseImage` fields, they are going to sit in `ClusterConfiguration` and using that one is a safe bet for me (as long as #66219 gets merged).